test/python/compiler/test_scheduler.py

```diff
@@ -0,0 +1,91 @@
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Scheduler Test."""
+
+import unittest
+
+from qiskit.circuit import QuantumRegister, ClassicalRegister, QuantumCircuit
+from qiskit.exceptions import QiskitError
+from qiskit.pulse import InstructionScheduleMap, Schedule
+from qiskit.test import QiskitTestCase
+from qiskit.test.mock import FakeOpenPulse3Q
+from qiskit.compiler.scheduler import schedule
+
+
+class TestCircuitScheduler(QiskitTestCase):
+    """Tests for scheduling."""
+
+    def setUp(self):
+        super().setUp()
+        qr = QuantumRegister(2, name="q")
+        cr = ClassicalRegister(2, name="c")
+        self.circ = QuantumCircuit(qr, cr, name="circ")
+        self.circ.cx(qr[0], qr[1])
+        self.circ.measure(qr, cr)
+
+        qr2 = QuantumRegister(2, name="q")
+        cr2 = ClassicalRegister(2, name="c")
+        self.circ2 = QuantumCircuit(qr2, cr2, name="circ2")
+        self.circ2.cx(qr2[0], qr2[1])
+        self.circ2.measure(qr2, cr2)
+
+        self.backend = FakeOpenPulse3Q()
+        self.backend_config = self.backend.configuration()
+        self.num_qubits = self.backend_config.n_qubits
+
+    def test_instruction_map_and_backend_not_supplied(self):
+        """Test instruction map and backend not supplied."""
+        with self.assertRaisesRegex(
+            QiskitError,
+            r"Must supply either a backend or InstructionScheduleMap for scheduling passes.",
+        ):
+            schedule(self.circ)
+
+    def test_instruction_map_and_backend_defaults_unavailable(self):
+        """Test backend defaults unavailable when backend is provided, but instruction map is not."""
+        self.backend._defaults = None
+        with self.assertRaisesRegex(
+            QiskitError, r"The backend defaults are unavailable. The backend may not support pulse."
+        ):
+            schedule(self.circ, self.backend)
+
+    def test_measurement_map_and_backend_not_supplied(self):
+        """Test measurement map and backend not supplied."""
+        with self.assertRaisesRegex(
+            QiskitError,
+            r"Must supply either a backend or a meas_map for scheduling passes.",
+        ):
+            schedule(self.circ, inst_map=InstructionScheduleMap())
+
+    def test_schedules_single_circuit(self):
+        """Test scheduling of a single circuit."""
+        circuit_schedule = schedule(self.circ, self.backend)
+
+        self.assertIsInstance(circuit_schedule, Schedule)
+        self.assertEqual(circuit_schedule.name, "circ")
+
+    def test_schedules_multiple_circuits(self):
+        """Test scheduling of multiple circuits."""
+        circuits = [self.circ, self.circ2]
+        circuit_schedules = schedule(circuits, self.backend)
```

Comment:
By default the `QISKIT_PARALLEL` env variable is set to `FALSE` in the CI jobs to avoid overloading the small CI VMs with too much parallel work (unit tests are executed in parallel). To ensure we're testing this with multiprocessing enabled so we're actually exercising the parallel dispatch it'd probably be good to something like:

```suggestion
        if sys.platform == "win32":
            PARALLEL_DEFAULT = False
        # On macOS default false on Python >=3.8
        elif sys.platform == "darwin":
            if sys.version_info[0] == 3 and sys.version_info[1] >= 8:
                parallel_default = False
            else:
                parallel_default = True
        # On linux (and other OSes) default to True
        else:
            parallel_default = True
        self.addCleanup(lambda val: os.environ["QISKIT_PARALLEL"] = val, os.getenv("QISKIT_PARALLEL", parallel_default)
        os.environ["QISKIT_PARALLEL"] = parallel_default
        circuit_schedules = schedule(circuits, self.backend)
```

(I haven't tested that so it might not work especially not without adding the missing imports)

That will override the set value of the `QISKIT_PARALLEL` env variable with the platform specific default and then restore it after the test. This way we should get full coverage for this change.

Reply:
@mtreinish Perhaps we could make a test helper function for this `QISKIT_PARALLEL` manipulation that can be used across multiple tests or files, if needed? That would make it so that we can simply add a single function call wherever it is needed now or in the future. If we think that's a good idea, where would be the best location to add that function?

Reply:
Yeah making this a standalone method is a good idea. You can honestly just do it in this class to start since nothing else will need it yet. Or also in the base test class definition is the other logical place to put it: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/test/base.py#L68

Reply:
For this change, I noticed that the same logic for determining the parallel default was implemented [here](https://github.com/Qiskit/qiskit-terra/blob/e5087294007550ac3c2a3691e510de204fc98fac/qiskit/tools/parallel.py#L66), so I moved that into a function that ensures we aren't duplicating this code. Having this encapsulated into a function also enabled me to add tests for our decision making regarding these defaults. All the changes found in [ac1e449](https://github.com/Qiskit/qiskit-terra/pull/8110/commits/ac1e4497843f82099b33bac9494b09175cda73cc) contribute to the addition of an `enableParallelProcessing` method on the base test class.